File: examples/c/tabletop_weather_station.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_lcd_128x64.h"
#include "bricklet_air_quality.h"

#define HOST "localhost"
#define PORT 4223

void millisleep(uint32_t msec); // defined in ip_connection.c

typedef struct {
	IPConnection ipcon;
	LCD128x64 lcd;
	bool lcd_created;
	AirQuality air_quality;
} TabletopWeatherStation;

void cb_all_values(int32_t iaq_index, uint8_t iaq_index_accuracy, int32_t temperature,
                   int32_t humidity, int32_t air_pressure, void *user_data) {
	TabletopWeatherStation *tws = (TabletopWeatherStation *)user_data;

	if(tws->lcd_created) {
		char text[30] = {'\0'};

		sprintf(text, "IAQ:      %6d", iaq_index);
		lcd_128x64_write_line(&tws->lcd, 2, 0, text);

		// 0xF8 == ° on LCD 128x64 charset
		sprintf(text, "Temp:     %6.2f %cC", temperature/100.0, 0xF8);
		lcd_128x64_write_line(&tws->lcd, 3, 0, text);
		sprintf(text, "Humidity: %6.2f %%RH", humidity/100.0);
		lcd_128x64_write_line(&tws->lcd, 4, 0, text);
		sprintf(text, "Air Pres: %6.2f hPa", air_pressure/100.0);
		lcd_128x64_write_line(&tws->lcd, 5, 0, text);
	}
}

void cb_connected(uint8_t connected_reason, void *user_data) {
	TabletopWeatherStation *tws = (TabletopWeatherStation *)user_data;

	// Eumerate again after auto-reconnect
	if(connected_reason == IPCON_CONNECT_REASON_AUTO_RECONNECT) {
		printf("Auto Reconnect\n");

		while(true) {
			int rc = ipcon_enumerate(&tws->ipcon);
			if(rc < 0) {
				fprintf(stderr, "Could not enumerate: %d\n", rc);
				millisleep(1000);
				continue;
			}
			break;
		}
	}
}

void cb_enumerate(const char *uid, const char *connected_uid,
                  char position, uint8_t hardware_version[3],
                  uint8_t firmware_version[3], uint16_t device_identifier,
                  uint8_t enumeration_type, void *user_data) {
	TabletopWeatherStation *tws = (TabletopWeatherStation *)user_data;
	int rc;

	// avoid unused parameter warning
	(void)connected_uid; (void)position; (void)hardware_version; (void)firmware_version;

	if(enumeration_type == IPCON_ENUMERATION_TYPE_CONNECTED ||
	   enumeration_type == IPCON_ENUMERATION_TYPE_AVAILABLE) {
		if(device_identifier == LCD_128X64_DEVICE_IDENTIFIER) {
			lcd_128x64_create(&tws->lcd, uid, &tws->ipcon);
			lcd_128x64_clear_display(&tws->lcd);
			lcd_128x64_write_line(&tws->lcd, 0, 0, "   Weather Station");
			tws->lcd_created = true;
			printf("LCD 128x64 initialized\n");
		} else if(device_identifier == AIR_QUALITY_DEVICE_IDENTIFIER) {
			air_quality_create(&tws->air_quality, uid, &tws->ipcon);
			air_quality_register_callback(&tws->air_quality,
			                              AIR_QUALITY_CALLBACK_ALL_VALUES,
			                              (void *)cb_all_values,
			                              (void *)tws);
			rc = air_quality_set_all_values_callback_configuration(&tws->air_quality, 1000, false);

			if(rc < 0) {
				fprintf(stderr, "Air Quality init failed: %d\n", rc);
			} else {
				printf("Air Quality initialized\n");
			}
		}
	}
}

int main() {
	TabletopWeatherStation tws;

	tws.lcd_created = false;
	ipcon_create(&tws.ipcon); // Create IP connection

	// Connect to brickd (retry if not possible)
	while(true) {
		int rc = ipcon_connect(&tws.ipcon, HOST, PORT);
		if(rc < 0) {
			fprintf(stderr, "Could not connect to brickd: %d\n", rc);
			millisleep(1000);
			continue;
		}
		break;
	}

	ipcon_register_callback(&tws.ipcon,
	                        IPCON_CALLBACK_ENUMERATE,
	                        (void *)cb_enumerate,
	                        (void *)&tws);

	ipcon_register_callback(&tws.ipcon,
	                        IPCON_CALLBACK_CONNECTED,
	                        (void *)cb_connected,
	                        (void *)&tws);

	// Enumerate Bricks and Bricklets (retry if not possible)
	while(true) {
		int rc = ipcon_enumerate(&tws.ipcon);
		if(rc < 0) {
			fprintf(stderr, "Could not enumerate: %d\n", rc);
			millisleep(1000);
			continue;
		}
		break;
	}

	printf("Press key to exit\n");
	getchar();
	ipcon_destroy(&tws.ipcon);
	return 0;
}

File: examples/php/TabletopWeatherStation.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletLCD128x64.php');
require_once('Tinkerforge/BrickletAirQuality.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletLCD128x64;
use Tinkerforge\BrickletAirQuality;

class WeatherStation
{
	const HOST = 'localhost';
	const PORT = 4223;

	public function __construct()
    {
		$this->brickletLCD = null;
		$this->brickletAirQuality = null;
		$this->ipcon = new IPConnection(); // Create IP connection

		// Connect to brickd (retry if not possible)
		while(true) {
			try {
				$this->ipcon->connect(self::HOST, self::PORT);
				break;
			} catch(Exception $e) {
				sleep(1);
			}
		}

		$this->ipcon->registerCallback(IPConnection::CALLBACK_ENUMERATE,
		                               array($this, 'cb_enumerate'));
		$this->ipcon->registerCallback(IPConnection::CALLBACK_CONNECTED,
		                               array($this, 'cb_connected'));

		// Enumerate Bricks and Bricklets (retry if not possible)
		while(true) {
			try {
				$this->ipcon->enumerate();
				break;
			} catch(Exception $e) {
				sleep(1);
			}
		}
	}

	function cb_all_values($iaq_index, $iaq_index_accuracy, $temperature, $humidity, $air_pressure)
	{
		if($this->brickletLCD != null) {
			$this->brickletLCD->writeLine(2, 0, sprintf("IAQ:      %6d", $iaq_index));
			// 0xF8 == ° on LCD 128x64 charset
			$this->brickletLCD->writeLine(3, 0, sprintf("Temp:     %6.2f %cC", $temperature/100.0, 0xF8));
			$this->brickletLCD->writeLine(4, 0, sprintf("Humidity: %6.2f %%RH", $humidity/100.0));
			$this->brickletLCD->writeLine(5, 0, sprintf("Air Pres: %6.2f hPa", $air_pressure/100.0));
		}
	}

	function cb_enumerate($uid, $connectedUid, $position, $hardwareVersion,
	                      $firmwareVersion, $deviceIdentifier, $enumerationType)
	{
		if($enumerationType == IPConnection::ENUMERATION_TYPE_CONNECTED ||
		   $enumerationType == IPConnection::ENUMERATION_TYPE_AVAILABLE) {
			if($deviceIdentifier == BrickletLCD128x64::DEVICE_IDENTIFIER) {
				try {
					// Initialize newly enumerated LCD128x64 Bricklet
					$this->brickletLCD = new BrickletLCD128x64($uid, $this->ipcon);
					$this->brickletLCD->clearDisplay();
					$this->brickletLCD->writeLine(0, 0, "   Weather Station");
					echo "LCD 128x64 initialized\n";
				} catch(Exception $e) {
					$this->brickletLCD = null;
					echo "LCD 128x64 init failed: $e\n";
				}
			} else if($deviceIdentifier == BrickletAirQuality::DEVICE_IDENTIFIER) {
				try {
					// Initialize newly enumaratedy Air Quality Bricklet and configure callbacks
					$this->brickletAirQuality = new BrickletAirQuality($uid, $this->ipcon);
					$this->brickletAirQuality->setAllValuesCallbackConfiguration(1000, true);
					$this->brickletAirQuality->registerCallback(BrickletAirQuality::CALLBACK_ALL_VALUES,
					                                            array($this, 'cb_all_values'));
					echo "Air Quality initialized\n";
				} catch(Exception $e) {
					$this->brickletAirQuality = null;
					echo "Air Quality init failed: $e\n";
				}
			} 
		}
	}

	function cb_connected($connectedReason)
	{
		// Eumerate again after auto-reconnect
		if($connectedReason == IPConnection::CONNECT_REASON_AUTO_RECONNECT) {
			echo "Auto Reconnect\n";

			while(true) {
				try {
					$this->ipcon->enumerate();
					break;
				} catch(Exception $e) {
					sleep(1);
				}
			}
		}
	}
}

$weatherStation = new WeatherStation();
echo "Press ctrl+c to exit\n";
$weatherStation->ipcon->dispatchCallbacks(-1);

?>
